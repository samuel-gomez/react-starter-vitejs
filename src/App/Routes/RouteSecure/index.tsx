export { default as RouteSecure, AuthorizedProfile } from './RouteSecure';
