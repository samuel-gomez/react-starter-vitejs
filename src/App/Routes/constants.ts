import { ROUTE_URL_HOME as HOME } from 'pages/Home/constants';
import { ROUTE_URL_NOTFOUND as NOTFOUND } from 'pages/NotFound/constants';
import { ROUTE_URL_UNAUTHORIZE as UNAUTHORIZE } from 'pages/Unauthorize/constants';

const ROUTE_URLS = {
  HOME,
  NOTFOUND,
  UNAUTHORIZE,
};

export default ROUTE_URLS;
