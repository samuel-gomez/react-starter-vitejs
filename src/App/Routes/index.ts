export { default, withAuth } from './Routes';
