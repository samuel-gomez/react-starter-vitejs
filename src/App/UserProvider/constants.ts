export const NON_CONNECTE = 'Non ConnectÃ©';

// Used when oidc configuration is disabled
const useOidcUserMock = () => ({
  oidcUser: {
    name: 'John Doe',
    given_name: 'John',
    family_name: 'Doe',
    email: 'john.doe@test.fr',
    email_verified: true,
    website: 'test.fr',
    sub: '2',
  },
});

export default useOidcUserMock;
