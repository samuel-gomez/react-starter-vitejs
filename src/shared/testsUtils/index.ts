export * from './customRender';
export { default as emptyFunction } from '../helpers/emptyFunction';
export { default as MOCK_API_URL } from './constants';
export { default as renderWithContainer } from './renderWithContainer';
