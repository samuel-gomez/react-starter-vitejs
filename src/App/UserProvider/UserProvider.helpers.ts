import { useOidcUser } from '@axa-fr/react-oidc';
import isEmpty from 'lodash/isEmpty';
import { PROFILS } from 'shared/constants';
import useOidcUserMock, { NON_CONNECTE } from './constants';

type ToidcUser = {
  member_of?: string[];
  name?: string;
  axa_uid_racf?: string;
};

/**
 * getAuthName
 */

type TgetAuthName = {
  oidcUser: ToidcUser;
  isEmptyFn?: typeof isEmpty;
};

export const getAuthName = ({ oidcUser, isEmptyFn = isEmpty }: TgetAuthName) => (!isEmptyFn(oidcUser?.name ?? '') ? oidcUser.name : NON_CONNECTE);

/**
 * setAuthRole
 */
type TsetAuthRole = {
  memberOf: string;
  profils?: typeof PROFILS;
};

export const setAuthRole = ({ memberOf, profils = PROFILS }: TsetAuthRole) =>
  profils.map(profil => (memberOf.search(`${profil}`) !== -1 ? profil : '')).join('') || memberOf;

/**
 * getAuthRole
 */

type TgetAuthRole = {
  oidcUser: ToidcUser;
  setAuthRoleFn?: typeof setAuthRole;
};

export const getAuthRole = ({ oidcUser, setAuthRoleFn = setAuthRole }: TgetAuthRole) =>
  setAuthRoleFn({ memberOf: oidcUser?.member_of?.[0].replace('CN=', '') ?? '' });

/**
 * getAuthUid
 */

type TgetAuthUid = {
  oidcUser: ToidcUser;
  isEmptyFn?: typeof isEmpty;
};

export const getAuthUid = ({ oidcUser }: TgetAuthUid) => oidcUser?.axa_uid_racf ?? '';

/**
 * extractDataFromOAuthToken
 * @param {Object} oidcUser
 */

type TextractDataFromOAuthToken = {
  oidcUser: ToidcUser;
  getAuthNameFn?: typeof getAuthName;
  getAuthRoleFn?: typeof getAuthRole;
  getAuthUidFn?: typeof getAuthUid;
};

export const extractDataFromOAuthToken = ({
  oidcUser,
  getAuthNameFn = getAuthName,
  getAuthRoleFn = getAuthRole,
  getAuthUidFn = getAuthUid,
}: TextractDataFromOAuthToken) => ({
  authName: getAuthNameFn({ oidcUser }),
  authRole: getAuthRoleFn({ oidcUser }),
  authUid: getAuthUidFn({ oidcUser }),
  isLoading: !oidcUser,
});

export const getUserInfos = ({
  isEnabled,
  useOidcUserFn = useOidcUser,
  useOidcUserMockFn = useOidcUserMock,
}: {
  isEnabled?: boolean;
  useOidcUserFn?: typeof useOidcUser;
  useOidcUserMockFn?: typeof useOidcUserMock;
}) => (isEnabled ? useOidcUserFn : useOidcUserMockFn);
