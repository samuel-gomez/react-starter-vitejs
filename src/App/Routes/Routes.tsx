/* eslint-disable max-lines-per-function */
import { useContext, ComponentProps } from 'react';
import { Routes, Route, Navigate } from 'react-router-dom';
import Home from 'pages/Home';
import PageNotFound from 'pages/NotFound';
import PageUnauthorize from 'pages/Unauthorize';
import ROUTE_URL from 'App/Routes/constants';
import { UserContext } from 'App/UserProvider';
import Loader, { MODES } from 'shared/components/Loader';
import { PROFILS } from 'shared/constants';

export const withAuth = <T extends object>(
  Component: React.ComponentType<T>,
  UserContextObj = UserContext,
  authorized = PROFILS,
  NavigateCmpt = Navigate,
  LoaderCmpt = Loader,
) => {
  const NewComp = (props: ComponentProps<typeof Component> | object) => {
    const userContext = useContext(UserContextObj);

    if (userContext.isEnabled && userContext.isLoading) {
      return <LoaderCmpt text="Chargement des donnÃ©es utilisateur..." mode={MODES.get} classModifier="fullscreen" />;
    }

    return authorized.includes(userContext?.authRole) || !userContext.isEnabled ? (
      <Component {...(props as T)} />
    ) : (
      <NavigateCmpt to={ROUTE_URL.UNAUTHORIZE} />
    );
  };

  return <NewComp />;
};

type TRoutesCmpt = {
  HomeCmpt?: typeof Home;
  PageUnauthorizeCmpt?: typeof PageUnauthorize;
  withAuthFn?: typeof withAuth;
};

const RoutesCmpt = ({ HomeCmpt = Home, PageUnauthorizeCmpt = PageUnauthorize, withAuthFn = withAuth }: TRoutesCmpt) => (
  <Routes>
    <Route path={ROUTE_URL.HOME} element={withAuthFn(HomeCmpt)} />
    <Route path={ROUTE_URL.UNAUTHORIZE} element={<PageUnauthorizeCmpt />} />
    <Route path="*" element={<PageNotFound />} />
  </Routes>
);

export default RoutesCmpt;
