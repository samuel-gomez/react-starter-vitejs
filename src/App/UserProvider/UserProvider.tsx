import { EnvironmentContext } from 'App/EnvironmentProvider';
import { createContext, useContext, useMemo, type ReactNode } from 'react';
import { extractDataFromOAuthToken, getUserInfos } from './UserProvider.helpers';

export type TUserContext = ReturnType<typeof extractDataFromOAuthToken> & {
  isLoading?: boolean;
};

export const UserContext = createContext<TUserContext>({
  authName: '',
  authRole: '',
  authUid: '',
  isLoading: true,
});
UserContext.displayName = 'UserContext';

type TUserProvider = {
  children: ReactNode;
  getUserInfosFn?: typeof getUserInfos;
  extractDataFromOAuthTokenFn?: typeof extractDataFromOAuthToken;
  isEnabled?: boolean;
};

const UserProvider = ({
  children,
  getUserInfosFn = getUserInfos,
  extractDataFromOAuthTokenFn = extractDataFromOAuthToken,
  ...rest
}: TUserProvider) => {
  const { environment } = useContext(EnvironmentContext);
  const { oidcUser } = getUserInfosFn({ isEnabled: environment?.oidc?.isEnabled })();

  const value = useMemo(() => ({ ...extractDataFromOAuthTokenFn({ oidcUser }), ...rest }), [extractDataFromOAuthTokenFn, oidcUser, rest]);
  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
};

export default UserProvider;
