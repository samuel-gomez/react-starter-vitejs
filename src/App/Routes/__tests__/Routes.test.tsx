import { describe, it, expect, vi } from 'vitest';
import { ComponentType, createContext } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import UserProvider from 'App/UserProvider';
import { renderWithWrapperStaticRouter } from 'shared/testsUtils';
import { TUserContext } from 'App/UserProvider/UserProvider';
import Routes, { withAuth } from '..';

const defaultPropsMock = {
  HomeCmpt: () => <div>HomeCmpt</div>,
  PageUnauthorizeCmpt: () => <div>PageUnauthorizeCmpt</div>,
};

type TrenderRoute = {
  role?: string;
  name?: string;
  route?: string;
  defaultProps?: object;
};

const renderRoute = ({ role = '', name = '', route = '/', defaultProps = defaultPropsMock }: TrenderRoute) => {
  const useOidcUser = vi.fn().mockReturnValue({
    oidcUser: { name, member_of: [`CN=${role}`] },
  });

  return render(
    <MemoryRouter initialEntries={[route]}>
      <UserProvider useOidcUserFn={useOidcUser} isEnabled>
        <Routes {...defaultProps} />
      </UserProvider>
    </MemoryRouter>,
  );
};

describe('<Routes />', () => {
  it.each`
    role  | name        | route
    ${''} | ${'Samuel'} | ${'/'}
    ${''} | ${'Samuel'} | ${'/forbidden'}
  `('Should render page when user profil is authorized, role: $role, name: $name, route: $route', ({ role, name, route }) => {
    const { asFragment } = renderRoute({ role, name, route });
    expect(asFragment()).toMatchSnapshot();
  });

  it.each`
    role                   | name     | route
    ${'user unauthorized'} | ${'sam'} | ${'/'}
  `('Should render forbidden page when user profil is unauthorized, role: $role, name: $name, route: $route', ({ role, name, route }) => {
    const UserContextObjMock = createContext<TUserContext>({ authName: '', authRole: role, authUid: '', isEnabled: true, isLoading: false });
    const { getByText } = renderRoute({
      role,
      name,
      route,
      defaultProps: { ...defaultPropsMock, withAuthFn: (...args: ComponentType<object>[]) => withAuth(args[0], UserContextObjMock, ['admin']) },
    });
    expect(getByText('PageUnauthorizeCmpt')).toBeInTheDocument();
  });

  it.each`
    route
    ${'/no-exist-route'}
  `('Should render 404 when election route: $route is not correct', ({ route }) => {
    const { getByText } = renderRoute({ route, role: '', name: 'samuel' });
    expect(getByText('404')).toBeInTheDocument();
  });
});

describe('Render withAuth', () => {
  const Component = () => <p>component</p>;
  it.each`
    authRole | authName
    ${''}    | ${'Samuel'}
  `('Should render RouteCmpt when user profile is authorized (authRole, authName)', ({ authRole, authName }) => {
    const UserContextObjMock = createContext<TUserContext>({ authRole, authName, authUid: '', isEnabled: true, isLoading: false });
    const { getByText } = renderWithWrapperStaticRouter(withAuth(Component, UserContextObjMock));
    expect(getByText('component')).toBeInTheDocument();
  });

  it('Should render NavigateCmpt when user profile is unauthorized', () => {
    const UserContextObjMock = createContext<TUserContext>({
      authName: '',
      authRole: 'unauthorized',
      authUid: '',
      isEnabled: true,
      isLoading: false,
    });
    const NavigateCmpt = vi.fn().mockReturnValue(null);
    renderWithWrapperStaticRouter(withAuth(Component, UserContextObjMock, [''], NavigateCmpt));
    expect(NavigateCmpt).toHaveBeenCalled();
  });

  it('Should render LoaderCmpt when user profile is loading', () => {
    const LoaderCmpt = vi.fn();
    const UserContextObjMock = createContext<TUserContext>({ authName: '', authRole: '', authUid: '', isEnabled: true, isLoading: true });
    const NavigateCmpt = vi.fn().mockReturnValue(null);
    renderWithWrapperStaticRouter(withAuth(Component, UserContextObjMock, [''], NavigateCmpt, LoaderCmpt));
    expect(LoaderCmpt).toHaveBeenCalled();
  });
});
