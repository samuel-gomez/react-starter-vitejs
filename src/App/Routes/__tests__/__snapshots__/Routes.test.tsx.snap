// Vitest Snapshot v1

exports[`<Routes /> > Should render page when user profil is authorized, role: '', name: 'Samuel', route: '/' 1`] = `
<DocumentFragment>
  <div>
    HomeCmpt
  </div>
</DocumentFragment>
`;

exports[`<Routes /> > Should render page when user profil is authorized, role: '', name: 'Samuel', route: '/forbidden' 1`] = `
<DocumentFragment>
  <div>
    PageUnauthorizeCmpt
  </div>
</DocumentFragment>
`;
